File: app-flutter/lib/core/services/popularity_feed_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:nebras_mobile_app/core/firebase/firestore_sync_config.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// يقرأ `aggregates_popular/top_weekly` — مستند واحد لتجنّب استعلامات باهظة.
///
/// كاش محلّي 6 ساعات. راجع [docs/CONTENT_ORDERING_AND_HOME.md].
class PopularityFeedService {
  PopularityFeedService._();

  static final PopularityFeedService instance = PopularityFeedService._();

  static const _docPath = 'aggregates_popular/top_weekly';
  static const _cacheKey = 'popular_weekly_ids_v1';
  static const _cacheAtKey = 'popular_weekly_cached_at_v1';
  static const _cacheTtl = Duration(hours: 6);

  List<String> _ids = const [];
  DateTime? _loadedAt;

  List<String> get popularIds => List.unmodifiable(_ids);

  Future<List<String>> loadPopularIds({bool forceRefresh = false}) async {
    if (!forceRefresh &&
        _loadedAt != null &&
        DateTime.now().difference(_loadedAt!) < _cacheTtl &&
        _ids.isNotEmpty) {
      return _ids;
    }

    final prefs = await SharedPreferences.getInstance();
    if (!forceRefresh) {
      final cachedAt = prefs.getInt(_cacheAtKey);
      final cachedRaw = prefs.getString(_cacheKey);
      if (cachedAt != null &&
          cachedRaw != null &&
          DateTime.now().millisecondsSinceEpoch - cachedAt <
              _cacheTtl.inMilliseconds) {
        try {
          final list = (jsonDecode(cachedRaw) as List)
              .map((e) => e.toString())
              .toList();
          _ids = list;
          _loadedAt = DateTime.now();
          return _ids;
        } catch (_) {}
      }
    }

    try {
      final snap = await FirestoreSyncConfig.instance.doc(_docPath).get();
      if (!snap.exists) {
        _ids = const [];
        _loadedAt = DateTime.now();
        return _ids;
      }
      final data = snap.data() ?? {};
      final rawIds = data['ids'];
      if (rawIds is List) {
        _ids = rawIds.map((e) => e.toString()).where((s) => s.isNotEmpty).toList();
      } else {
        _ids = const [];
      }
      _loadedAt = DateTime.now();
      await prefs.setString(_cacheKey, jsonEncode(_ids));
      await prefs.setInt(_cacheAtKey, DateTime.now().millisecondsSinceEpoch);
    } catch (e) {
      debugPrint('[PopularityFeed] load failed: $e');
    }
    return _ids;
  }

  void clearMemoryCache() {
    _ids = const [];
    _loadedAt = null;
  }

  Future<void> clearDiskCache() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_cacheKey);
    await prefs.remove(_cacheAtKey);
    clearMemoryCache();
  }
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/service/PopularityFeedService.kt
package com.nebras.mobile.core.service

import android.util.Log
import com.nebras.mobile.core.data.LocalStore
import com.nebras.mobile.core.firebase.FirestoreSyncConfig
import kotlinx.coroutines.tasks.await
import org.json.JSONArray

/**
 * يقرأ `aggregates_popular/top_weekly` — مستند واحد لتجنّب استعلامات باهظة.
 * كاش محلّي 6 ساعات.
 */
object PopularityFeedService {

    private const val TAG = "PopularityFeed"
    private const val DOC_PATH = "aggregates_popular/top_weekly"
    private const val CACHE_KEY = "popular_weekly_ids_v1"
    private const val CACHE_AT_KEY = "popular_weekly_cached_at_v1"
    private const val CACHE_TTL_MILLIS = 6L * 60 * 60 * 1000

    private lateinit var store: LocalStore

    @Volatile
    private var ids: List<String> = emptyList()

    @Volatile
    private var loadedAt: Long? = null

    fun init(localStore: LocalStore) {
        store = localStore
    }

    val popularIds: List<String>
        get() = ids

    suspend fun loadPopularIds(forceRefresh: Boolean = false): List<String> {
        val now = System.currentTimeMillis()
        val cachedAtMemory = loadedAt
        if (!forceRefresh &&
            cachedAtMemory != null &&
            now - cachedAtMemory < CACHE_TTL_MILLIS &&
            ids.isNotEmpty()
        ) {
            return ids
        }

        if (!forceRefresh) {
            val cachedAt = store.getLong(CACHE_AT_KEY, 0L)
            val cachedRaw = store.getJsonArray(CACHE_KEY)
            if (cachedAt > 0 && cachedRaw != null && now - cachedAt < CACHE_TTL_MILLIS) {
                val list = cachedRaw.toStringList()
                ids = list
                loadedAt = now
                return ids
            }
        }

        runCatching {
            val snapshot = FirestoreSyncConfig.instance.document(DOC_PATH).get().await()
            if (!snapshot.exists()) {
                ids = emptyList()
                loadedAt = now
                return ids
            }
            val rawIds = snapshot.get("ids")
            ids = if (rawIds is List<*>) {
                rawIds.mapNotNull { it?.toString()?.takeIf(String::isNotEmpty) }
            } else {
                emptyList()
            }
            loadedAt = now
            val array = JSONArray()
            ids.forEach { array.put(it) }
            store.putJsonArray(CACHE_KEY, array)
            store.putLong(CACHE_AT_KEY, now)
        }.onFailure {
            Log.d(TAG, "load failed: ${it.message}")
        }

        return ids
    }

    fun clearMemoryCache() {
        ids = emptyList()
        loadedAt = null
    }

    fun clearDiskCache() {
        store.remove(CACHE_KEY)
        store.remove(CACHE_AT_KEY)
        clearMemoryCache()
    }
}

internal fun JSONArray.toStringList(): List<String> {
    val out = ArrayList<String>(length())
    for (i in 0 until length()) {
        opt(i)?.toString()?.takeIf(String::isNotEmpty)?.let(out::add)
    }
    return out
}
